compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AnnotationLowering.kt

```diff
@@ -5,26 +5,181 @@
+import org.jetbrains.kotlin.backend.jvm.intrinsics.IrIntrinsicMethods
+import org.jetbrains.kotlin.backend.jvm.intrinsics.KClassJavaProperty
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.builders.declarations.buildField
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.builders.irGet
 import org.jetbrains.kotlin.ir.declarations.IrClass
 import org.jetbrains.kotlin.ir.declarations.IrConstructor
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.impl.IrGetterCallImpl
+import org.jetbrains.kotlin.ir.symbols.IrClassSymbol
+import org.jetbrains.kotlin.ir.symbols.IrSimpleFunctionSymbol
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.types.impl.makeTypeProjection
+import org.jetbrains.kotlin.ir.util.functions
 import org.jetbrains.kotlin.ir.util.isAnnotationClass
+import org.jetbrains.kotlin.ir.util.isKClass
+import org.jetbrains.kotlin.ir.util.isNonPrimitiveArray
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.types.Variance
 
 internal val annotationPhase = makeIrFilePhase(
     ::AnnotationLowering,
     name = "Annotation",
-    description = "Remove constructors from annotation classes"
+    description = "Remove constructors and modify field types in annotation classes"
 )
 
-private class AnnotationLowering() : ClassLoweringPass {
-    constructor(@Suppress("UNUSED_PARAMETER") context: BackendContext) : this()
+/**
+ * Remove the constructors from annotation classes and change the types of KClass
+ * and Array<KClass> fields to use java.lang.Class instead. This phase also rewrites
+ * the uses of annotation class fields appropriately.
+ */
+private class AnnotationLowering(private val context: JvmBackendContext) : FileLoweringPass, IrElementTransformerVoid() {
+
+    override fun lower(irFile: IrFile) {
+        irFile.transformChildrenVoid(this)
+    }
+
+    override fun visitClass(irClass: IrClass): IrStatement {
+        if (!irClass.isAnnotationClass) return super.visitClass(irClass)
 
-    override fun lower(irClass: IrClass) {
-        if (!irClass.isAnnotationClass) return
         irClass.declarations.removeIf {
             it is IrConstructor
         }
+
+        for (declaration in irClass.declarations)
+            if (declaration is IrSimpleFunction)
+                lowerAnnotationField(declaration)
+
+        return irClass
+    }
+
+    // Lower the types on annotation class fields (KClass -> Class, Array<KClass> -> Array<Class>)
+    private fun lowerAnnotationField(declaration: IrSimpleFunction) {
+        val property = declaration.correspondingProperty ?: return
```

Comment:
val property = declaration.correspondingPropertySymbol?.owner ?: return

Reply:
In my defense, I wrote the code before it was deprecated. :)
Anyways, changed it to use correspondingPropertySymbol.